crates/astria-gossipnet/src/network.rs

```diff
@@ -131,43 +136,119 @@ impl Network {
+            gossipsub::MessageAuthenticity::Signed(keypair),
             gossipsub_config,
         )
         .map_err(|e| eyre!("failed to create gossipsub behaviour: {}", e))?;
 
         let mut swarm = {
-            #[cfg(feature = "mdns")]
-            let mdns = mdns::tokio::Behaviour::new(mdns::Config::default(), local_peer_id)?;
+            let kademlia = if self.with_kademlia {
+                let mut cfg = KademliaConfig::default();
+                cfg.set_query_timeout(Duration::from_secs(5 * 60));
+                let store = MemoryStore::new(local_peer_id);
+                Some(Kademlia::with_config(local_peer_id, store, cfg))
+            } else {
+                None
+            };
+
+            let mdns = if self.with_mdns {
+                Some(mdns::tokio::Behaviour::new(
+                    mdns::Config::default(),
+                    local_peer_id,
+                )?)
+            } else {
+                None
+            };
+
             let behaviour = GossipnetBehaviour {
+                identify: identify::Behaviour::new(identify::Config::new(
+                    GOSSIPNET_PROTOCOL_ID.into(),
+                    public_key,
+                )),
                 gossipsub,
-                #[cfg(feature = "mdns")]
-                mdns,
                 ping: ping::Behaviour::default(),
+                mdns: mdns.into(),
+                kademlia: kademlia.into(),
             };
             SwarmBuilder::with_tokio_executor(transport, behaviour, local_peer_id).build()
         };
 
-        let listen_addr = format!("/ip4/0.0.0.0/tcp/{}", port);
+        let listen_addr = format!("/ip4/0.0.0.0/tcp/{}", self.port);
         swarm.listen_on(listen_addr.parse()?)?;
 
-        if let Some(addrs) = bootnodes {
+        if let Some(addrs) = self.bootnodes {
             addrs.iter().try_for_each(|addr| -> Result<_> {
-                debug!("dialing {:?}", addr);
-                let remote: Multiaddr = addr.parse()?;
-                swarm.dial(remote)?;
-                debug!("dialed {addr}");
+                let mut maddr: Multiaddr = addr.parse()?;
+                swarm.dial(maddr.clone())?;
+
+                let Some(peer_id) = maddr.pop() else {
+                    return Err(eyre!("failed to parse peer id from addr: {}", addr));
+                };
+
+                match peer_id {
+                    Protocol::P2p(peer_id) => {
+                        let peer_id = match PeerId::from_multihash(peer_id) {
+                            Ok(peer_id) => peer_id,
+                            Err(e) => {
+                                return Err(eyre!("failed to parse peer id from addr: {:?}", e));
+                            }
+                        };
+
+                        if let Some(kademlia) = swarm.behaviour_mut().kademlia.as_mut() {
+                            kademlia.add_address(&peer_id, maddr);
+                        }
+                    }
+                    _ => {
```

Comment:
IMO this is better to guard against refactoring:
```suggestion
                    other => {
```